File: runtime-ios/change-basemap-of-webmap/ViewController.m
//
//  ViewController.m
//
//  Created by Garima Dhakal on 7/31/14. Thanks to Nathan and Feng.
//

/* Implementation file
 * Implements switchBaseMapOnMapView method of AGSWebMap class.
 */

#import "ViewController.h"

static NSString* const kPublicWebmapId = @"8a567ebac15748d39a747649a2e86cf4";

@interface ViewController ()
@property (strong, nonatomic) AGSWebMap* webMap;
@property(nonatomic, strong) NSString* webmapId;
@end

@implementation ViewController

- (void)viewDidLoad{
    [super viewDidLoad];
    
    self.webmapId = kPublicWebmapId;
    
	//Create a webmap using the ID
    self.webMap = [AGSWebMap webMapWithItemId:self.webmapId credential:nil];
    
    //Open the webmap
    [self.webMap openIntoMapView:self.mapView];
    
    //Set the webmap's delegate
    self.webMap.delegate = self;
}

//Action for the button click event. It changes the basemap of the webmap.
- (IBAction)changeBasemap:(id)sender{
    
    //Store JSON in string format
    NSString* myCustomJson = @"{\"operationalLayers\":[],\"baseMap\":{\"baseMapLayers\":[{\"id\":\"defaultBasemap\",\"opacity\":1,\"visibility\":true,\"url\":\"http://services.arcgisonline.com/ArcGIS/rest/services/World_Topo_Map/MapServer\"}],\"title\":\"Topographic\"},\"version\":\"1.6\"}\"";
    
    //Convert string to data
    NSData* data = [myCustomJson dataUsingEncoding:NSUTF8StringEncoding];
    data = [data subdataWithRange:NSMakeRange(0, [data length]-1)];
    
    /* Instead, use JSON from URL
     
     NSURL* urlOfJson = [NSURL URLWithString: @"http://www.arcgis.com/sharing/content/items/0493b877e54c4308adc038191b0a85d9/data?f=pjson"];
     NSData* data = [NSData dataWithContentsOfURL:urlOfJson];
     
     */
    
    //use NSJSONSerialization
    NSError* error;
    NSDictionary* dictionaryFromJson = [NSJSONSerialization JSONObjectWithData:data options:kNilOptions error:&error];
    
    //Get the JSON value for the baseMap key
    NSDictionary* baseMapJson = [dictionaryFromJson objectForKey:@"baseMap"];
    
    //Create an instance of AGSWebMapBaseMap
    AGSWebMapBaseMap* newWebmapBasemap = [[AGSWebMapBaseMap alloc]initWithJSON:baseMapJson];
    
    //Switch the basemap of webmap. The webmap should be completely loaded before this method is called.
    [self.webMap switchBaseMapOnMapView:newWebmapBasemap];
    
    //Hide the switch button
    [self.changeBasemapBTN setHidden:YES];
}


#pragma mark - AGSWebMapDelegate method
- (void) webMapDidLoad:(AGSWebMap *)webMap{
    NSLog(@"Webmap is successfully loaded.");
}

#pragma mark - AGSWebMapDelegate method
-(void)webMap:(AGSWebMap *)webMap didFailToLoadWithError:(NSError *)error{
    
    NSLog(@"Error occurred while loading the webmap: %@", [error localizedDescription]);
}

#pragma mark - AGSWebMapDelegate method
-(void)webMap:(AGSWebMap*)webMap didSwitchBaseMap:(AGSWebMapBaseMap*)baseMap onMapView:(AGSMapView*)mapView{
    NSLog(@"Tried to load basemap layer and add it to the map.");
}

#pragma mark - AGSWebMapDelegate method
-(void) webMap:(AGSWebMap *) webMap didFailToLoadLayer:(AGSWebMapLayerInfo *)layerInfo baseLayer:(BOOL)baseLayer federated:(BOOL)federated withError:(NSError *)error{
 
    NSLog(@"Error while loading layer: %@",[error localizedDescription]);
}
@end

File: runtime-ios/10.2.x/AGSWebMap-CurrentLocation-Swift/ViewController.swift
//
//  ViewController.swift
//  MapViewDemo-Swift
//  Nathan Wu

import UIKit
import ArcGIS

class ViewController: UIViewController , AGSWebMapDelegate, AGSMapViewLayerDelegate {
    
    @IBOutlet weak var switchBtn: UISwitch!
    @IBOutlet weak var mapView: AGSMapView!
    var currentLocation: AGSLocation!
    let webmap = AGSWebMap(itemId: "f553d4b86e354c05917ad269de26f6f0", credential: nil)
    let switchBasemapWebmap = AGSWebMap(itemId: "d94dcdbe78e141c2b2d3a91d5ca8b9c9", credential: nil)
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.webmap.delegate = self
        self.switchBtn.setOn(false, animated: false)
        
        self.webmap.openIntoMapView(self.mapView)
        self.mapView.locationDisplay.startDataSource()
        
    }
    
    func webMapDidLoad(webMap: AGSWebMap!) {
        
        print("Load webmap!")
    }
    
    func webMap(webMap: AGSWebMap!, didLoadLayer layer: AGSLayer!) {
        
        currentLocation = self.mapView.locationDisplay.location
        print(currentLocation)
        
        let projectedMapPoint = AGSGeometryEngine.defaultGeometryEngine().projectGeometry(currentLocation.point, toSpatialReference: AGSSpatialReference.webMercatorSpatialReference()) as! AGSPoint;
        self.webmap.mapView.centerAtPoint(projectedMapPoint, animated: true)
        
    }
    @IBAction func switchBasemap(sender: UISwitch) {
        if switchBtn.on {
            print("Button is on")
            self.webmap.switchBaseMapOnMapView(self.switchBasemapWebmap.baseMap!)
        } else {
            print("Button is off")
            self.webmap.switchBaseMapOnMapView(self.webmap.baseMap!)
        }
    }

    func webMap(webMap: AGSWebMap!, didSwitchBaseMap baseMap: AGSWebMapBaseMap!, onMapView mapView: AGSMapView!) {
        print("Basemap did switch : \(self.webmap.baseMap.title)")
    }
    
    func webMap(webMap: AGSWebMap!, didFailToLoadWithError error: NSError!) {
        print("Error while loading webmap: \(error.localizedDescription)")
    }
    
}